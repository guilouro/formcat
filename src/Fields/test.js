import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import { Form } from '..'
import {
  InputField,
  SelectField,
  Submit,
  TextField,
  RadiosField,
  CheckboxField
} from '.'

jest.mock('lodash/debounce', () =>
  jest.fn(fn => {
    fn.cancel = jest.fn()
    return fn
  })
)

const mockResponse = {
  data: {
    first_name: 'Guilherme Louro',
    gender: 'Male',
    time_as_developer: '3-5',
    agree: true,
    message: 'I hope that these tests have been passed'
  },
  error: false,
  field: {
    agree: { checked: true, required: true, value: true },
    first_name: { checked: false, required: false, value: 'Guilherme Louro' },
    gender: { checked: false, required: false, value: 'Male' },
    time_as_developer: { checked: false, required: false, value: '3-5' },
    message: {
      checked: false,
      required: false,
      value: 'I hope that these tests have been passed'
    }
  }
}

describe('Fields', () => {
  it('Should submit correctly', done => {
    const onSubmit = jest.fn()
    const { container, getByText } = render(
      <Form onSubmit={onSubmit}>
        <InputField label="First Name" name="first_name" />
        <SelectField
          label="Gender"
          name="gender"
          options={[
            { label: 'Female', value: 'Female' },
            { label: 'Male', value: 'Male' }
          ]}
        />

        <RadiosField
          label="How many time do you work as developer?"
          name="time_as_developer"
          options={[
            { label: '1 - 3 years', value: '1-3' },
            { label: '3 - 5 years', value: '3-5' },
            { label: '+5 years', value: '+5' }
          ]}
        />

        <CheckboxField label="I agree" name="agree" required />

        <TextField label="Some message" name="message" />

        <Submit>Submit</Submit>
      </Form>
    )

    fireEvent.change(container.querySelector('input[name="first_name"]'), {
      target: { value: mockResponse.field.first_name.value }
    })

    fireEvent.change(container.querySelector('select'), {
      target: { value: mockResponse.field.gender.value }
    })

    fireEvent.click(
      container.querySelectorAll('input[name="time_as_developer"]')[1]
    )

    fireEvent.click(container.querySelector('input[name="agree"]'))

    fireEvent.change(container.querySelector('textarea'), {
      target: { value: mockResponse.field.message.value }
    })

    fireEvent.click(getByText(/Submit/g))

    setImmediate(() => {
      expect(onSubmit).toBeCalledWith(mockResponse)
      expect(container).toMatchSnapshot()
      done()
    })
  })
})
