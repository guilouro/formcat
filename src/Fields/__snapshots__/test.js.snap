// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields Should submit correctly 1`] = `
<div>
  <form
    novalidate=""
  >
    <div>
      <label
        for="First Name"
      >
        First Name
      </label>
      <input
        id="First Name"
        name="first_name"
        value="Guilherme Louro"
      />
    </div>
    <div>
      <label
        for="Gender"
      >
        Gender
      </label>
      <select
        id="Gender"
        name="gender"
      >
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Male"
        >
          Male
        </option>
      </select>
    </div>
    <div>
      How many time do you work as developer?
    </div>
    <div>
      <input
        id="1 - 3 years"
        name="time_as_developer"
        type="radio"
        value="1-3"
      />
      <label
        for="1 - 3 years"
      >
        1 - 3 years
      </label>
    </div>
    <div>
      <input
        id="3 - 5 years"
        name="time_as_developer"
        type="radio"
        value="3-5"
      />
      <label
        for="3 - 5 years"
      >
        3 - 5 years
      </label>
    </div>
    <div>
      <input
        id="+5 years"
        name="time_as_developer"
        type="radio"
        value="+5"
      />
      <label
        for="+5 years"
      >
        +5 years
      </label>
    </div>
    <div>
      <input
        id="I agree"
        name="agree"
        required=""
        type="checkbox"
        value="true"
      />
      <label
        for="I agree"
      >
        I agree
        *
      </label>
    </div>
    <div>
      <label
        for="Some message"
      >
        Some message
      </label>
      <textarea
        id="Some message"
        name="message"
      >
        I hope that these tests have been passed
      </textarea>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
